codersdk/workspacesdk/connector_internal_test.go

```diff
@@ -142,6 +146,218 @@ func TestTailnetAPIConnector_UplevelVersion(t *testing.T) {
+		// behavior matches the actual client coordinate route.
+		var (
+			peerID      = uuid.New()
+			resumeToken = r.URL.Query().Get("resume_token")
+		)
+		t.Logf("received resume token: %s", resumeToken)
+		assert.Equal(t, expectResumeToken, resumeToken)
+		if resumeToken != "" {
+			peerID, err = resumeTokenProvider.VerifyResumeToken(resumeToken)
+			assert.NoError(t, err, "failed to parse resume token")
+			if err != nil {
+				httpapi.Write(ctx, w, http.StatusUnauthorized, codersdk.Response{
+					Message: CoordinateAPIInvalidResumeToken,
+					Detail:  err.Error(),
+					Validations: []codersdk.ValidationError{
+						{Field: "resume_token", Detail: CoordinateAPIInvalidResumeToken},
+					},
+				})
+				return
+			}
+		}
+
+		sws, err := websocket.Accept(w, r, nil)
+		if !assert.NoError(t, err) {
+			return
+		}
+		testutil.RequireSendCtx(ctx, t, websocketConnCh, sws)
+		ctx, nc := codersdk.WebsocketNetConn(r.Context(), sws, websocket.MessageBinary)
+		err = svc.ServeConnV2(ctx, nc, tailnet.StreamID{
+			Name: "client",
+			ID:   peerID,
+			Auth: tailnet.ClientCoordinateeAuth{AgentID: agentID},
+		})
+		assert.NoError(t, err)
+	}))
+
+	fConn := newFakeTailnetConn()
+
+	clock := quartz.NewMock(t)
+	newTickerTrap := clock.Trap().NewTicker("tailnetAPIConnector", "refreshToken")
+	tickerResetTrap := clock.Trap().TickerReset("tailnetAPIConnector", "refreshToken", "reset")
+	defer newTickerTrap.Close()
+	uut := newTailnetAPIConnector(ctx, logger, agentID, svr.URL, clock, &websocket.DialOptions{})
+	uut.runConnector(fConn)
+
+	// Fetch first token.
+	trappedTicker := newTickerTrap.MustWait(ctx)
+	trappedTicker.Release()
+	waiter := clock.Advance(trappedTicker.Duration)
+	waiter.MustWait(ctx)
+	// We call ticker.Reset after each token fetch to apply the refresh duration
+	// requested by the server.
+	trappedReset := tickerResetTrap.MustWait(ctx)
+	trappedReset.Release()
+	require.NotNil(t, uut.resumeToken)
+	originalResumeToken := uut.resumeToken.Token
+
+	// Fetch second token.
+	waiter = clock.Advance(trappedReset.Duration)
+	waiter.MustWait(ctx)
+	trappedReset = tickerResetTrap.MustWait(ctx)
+	trappedReset.Release()
+	require.NotNil(t, uut.resumeToken)
+	require.NotEqual(t, originalResumeToken, uut.resumeToken.Token)
+	expectResumeToken = uut.resumeToken.Token
+	t.Logf("expecting resume token: %s", expectResumeToken)
+
+	// Sever the connection and expect it to reconnect with the resume token.
+	wsConn := testutil.RequireRecvCtx(ctx, t, websocketConnCh)
+	_ = wsConn.Close(websocket.StatusGoingAway, "test")
+
+	// Wait for the resume token to be refreshed.
+	trappedTicker = newTickerTrap.MustWait(ctx)
+	trappedTicker.Release()
+	waiter = clock.Advance(trappedTicker.Duration)
+	waiter.MustWait(ctx)
+	trappedReset = tickerResetTrap.MustWait(ctx)
+	trappedReset.Release()
+
+	// The resume token should have changed again.
+	require.NotNil(t, uut.resumeToken)
+	require.NotEqual(t, expectResumeToken, uut.resumeToken.Token)
```

Comment:
This could be a racy assertion, especially once the JWT conforms to the RFC where `exp` only has 1 second resolution. `tailnet.NewInsecureTestResumeTokenProvider()` uses the real clock, and 1 second may not have elapsed between the previous token and this one.

Here it seems that asserting the token is different is a proxy for asserting that we got a Refresh RPC call --- but trapping the Reset call above ensures that from a timing perspective.  If you really want to assert they're different, pass the Mock Clock to the ResumeTokenProvider so that the clock it uses is advancing during the test.

Reply:
This also affects the workspaceagents.go test so I fixed it there too.